### PR TITLE
[RFC] Add option to store the parameters during the optimization

### PR DESCRIPTION
**Pull request by PhilipVinc**
Add an option to JsonLog to store the parameters during the optimisation.
It will save them at every iteration when data is logged, inside a tar file with no compression.
(In the future we might activate compression but it is tricky)

Adds two experimental functions in netket.variational.experimental to read data from a file or a tar file.

I'd like opinions.

### CODE REVIEW

**Comment by PhilipVinc** [comment]
The name `variables` instead of `parameters` is because it is actually storing variables (the whole model state + parameters), and helps the user understand that he wants to access `vstate.variables` instead of `vstate.parameters`.

Also, it might make sense to break this functionality out of JsonLog and create a TarLog.

**Comment by PhilipVinc** [comment]
One advantage of having it in JsonLog is that the various files stored inside of the tar archive naturally correspond to the timesteps stored in the json file.
If we decouple the two, by creating a TarLog, we will need some way to communicate how to reconstruct the iteration number. Either by adding a file with all iterations inside the tar file(but that is hard to do, becauseyou cannot edit the content of a file in an archive) or by creating another file

**Comment by femtobit** [comment]
Hm, I feel we are starting to venture into the territory of inventing our own ad-hoc data formats.

I'm happy with the use of JSON for our default output since it's such a common format. Having a quick way to write the state at every iteration to a file sounds fine to me as well, but going beyond that, i.e., by introducing tar archives with compression etc., is the part where I feel we should try to not reinvent the wheel.

If we do want to have "all batteries included" logging we should maybe look into doing this in a somewhat more standardized and feature-rich format.
 
I've used [`zarr`](https://zarr.readthedocs.io/en/stable/) for simulation output in the past and liked it a lot. Some benefits:
1. It can efficiently store arrays, supporting the transparent use of compression algorithms (which can help a lot with some more regular array data such as `iters`) and saving to chunked files on disk (it also supports storing data in ZIP files, or even databases and other backends if needed).
2. It allows to have a hierarchy of groups in one dataset, similar to HDF5 and other formats.
3. It supports storing python `object`s via `msgpack`, among other formats, which should be helpful for some of our NetKet data.
4. Custom metadata can be added to arrays and groups which can come in handy sometimes.

One downside is that it is primarily supported in Python, but OTOH if we output some custom-format tar file that's also something people would need at least a bit of dedicated code to work with and NetKet is a Python framework, after all.

**Comment by PhilipVinc** [comment]
I don't want to include too many batteries in NetKet.
However I think that being able to log the parameters at every iteration is kind of important (you need it if you do time-evolution).

I picked a tar archive simply because it's... easy and universally supported:
 - You can untar it anytime without installing anything (tar is supported in Mac/win/linux out of the box) and get all .mpack objects
 - It's fast, and equivalent to just writing files.
 - If you run on a cluster where you have a limit to the number of files you can store (at least, that's my experience) you don't have issues with creating too many files.

> I'm happy with the use of JSON for our default output since it's such a common format

I considered dumping the binary string with the content of .mpack inside the json. I did not go that way because we are rewriting the .json file at every step (we cannot append to it). Though we could support it in principle...

> I've used zarr for simulation output in the past and liked it a lot. Some benefits:

I'll give a look. However a format that is somewhat widely supported would be handy. We can even support both, maybe? 

> One downside is that it is primarily supported in Python

I do my analysis in Julia, so... well. I can use Zarr from PyCall, but that's a bit ugly. Is it widely used?

**Comment by femtobit** [comment]
Well, there is https://github.com/meggart/Zarr.jl, though I am not sure how widely it is used.

I can live with the tar file for storing the time evolved parameters; but I think this may get us started down a rabbit hole: If we store states in our "tarred set of mpack" format, why not add more information like, say, energy gradients or the QFM (both would be very useful when debugging time evolution)? And soon we will have a whole ad-hoc "file-system in a tar archive" output format that we'll have to support for all times. And if we're going in that direction, why not build on some existing backend that can already support compression, different archive formats, etc. (I mentioned zarr to be specific and because I've like it when using it before, but there are alternatives.)

**Comment by PhilipVinc** [comment]
Nice! So I have nothing against Zarr. It also seems it's using a simpler reimplementation of hdf5 under the hood?

>Yes, this is the main reason I'd also not add too much data to the JSON. (Though we could optimize that. NetKet v2 did not rewrite the full log at every step as far as I remember.)

You kept the string in memory, but you had to rewrite the whole file at every iteration.
I have been benchmarking serialization/deserialization of JSOn and Tar and they are really not the bottleneck even with  networks with ~5000 parameters.


I think that we can have both? A TarLog.jl only logging the states, and if someone (you?) will ever contribute a ZarrLog, we'll have that one too... Also because deserialising that one is going to be a bit more involved, I guess...

There's a bunch of people that asked me about a way to store parameters during the evolution so I think we need a way to do that...

**Comment by femtobit** [comment]
> You kept the string in memory, but you had to rewrite the whole file at every iteration.

Actually, I did write some code to write the JSON line-by-line (but still keeping it as one valid JSON file) in v2:

https://github.com/netket/netket/blob/d2aa17adecfcbafe9568b8517caf9e2981745d48/Sources/Output/json_output_writer.hpp#L69-L88

Though if, as you say, the runtime and memory overhead of rewriting the full file is negligible it should be fine (and we could always optimize it in v3 in the same way if that becomes necessary).

> I think that we can have both? A TarLog.jl only logging the states, and if someone (you?) will ever contribute a ZarrLog, we'll have that one too... Also because deserialising that one is going to be a bit more involved, I guess...

Yes, alright, as a quick and easy way to store all parameters in the tar should be fine; though I am still pretty sure we will have people asking to add all kind of data to the tar over time so we should make sure this is easily possible or we have a good alternative.

> There's a bunch of people that asked me about a way to store parameters during the evolution so I think we need a way to do that...

Maybe it would also be good to have a way to combine `.run` and `.iter` on the driver? If one could do this:
```python
for i in driver.run("normal_json_log", ...):
    with open(f"states/{i}.mpack", "w") as file:
        file.write(flax.serialization.to_bytes(driver.state))
```
I think the need for adding all kinds of features to the JSON log would at least be reduced.

**Comment by PhilipVinc** [comment]
>Actually, I did write some code to write the JSON line-by-line (but still keeping it as one valid JSON file) in v2 (not saying this is particularly elegant, but it did the job 😅):

Ah, indeed. I thought you where referring to the python v2 (or what I sometimes call 2.9... the never released intermediate netket).

> Yes, alright, as a quick and easy way to store all parameters in the tar should be fine; though I am still pretty sure we will have people asking to add all kind of data to the tar over time so we should make sure this is easily possible or we have a good alternative.

This won't be possible. I'd rather have a good alternative. You seem to be the zar of the Zarr so 😉.

> Maybe it would also be good to have a way to combine .run and .iter on the driver? If one could do this:

You can kind of achieve this with callbacks.
If you know a way to support this, it would be good (but I don't know any)...

Maybe we should make `iter` internal and expose an iter that does what you propose?

File: netket/logging/tar_log.py
Comment (femtobit) [inline netket/logging/tar_log.py:53]: Is this save in case the simulation run crashes? Being able to recover from interruptions seems like an important use case for storing intermediate states on disk. If an improper shutdown of the process can leave the tar file in a corrupted state this would not work.

(Also, being able to already access the intermediate states in a long running simulation can be pretty useful by itself.)

File: netket/logging/tar_log.py
Comment (femtobit) [inline netket/logging/tar_log.py:42]: Since this only logs the variational parameters, maybe it should be called `TarStateLog` or similar?

Comment (PhilipVinc) [reply netket/logging/tar_log.py:42]: ok

File: netket/logging/tar_log.py
Comment (PhilipVinc) [reply netket/logging/tar_log.py:53]: This does close the file handle if an exception is thrown (it doesn't if there is a segmentation fault, but in that case it's impossible to recover).

However, to make him 'flush' the file I would have to open and close the file handle after every write. Closing is cheap (he just writes a small ending) but reopening is more expensive as he repress the index of thins inside the tar file.

I think it's not worth doing it.

Comment (femtobit) [reply netket/logging/tar_log.py:53]: Hm, in that case I'd prefer an option of just storing the files one by one in a directory (where this is not an issue as a crash would only affect the last file if one is unlucky).

But that's maybe separate from this PR.

Comment (PhilipVinc) [reply netket/logging/tar_log.py:53]: maybe we can rename to StateLog and have an option `tar=True/False`?

Comment (femtobit) [reply netket/logging/tar_log.py:53]: I think that would be pretty useful!

**Comment by PhilipVinc** [comment]
Ok, It's renamed to `StateLog` and takes a flag `tar=True/False` default false.
It puts everything in a folder. It deletes everything .mpack inside the folder

File: netket/logging/state_log.py
Comment (femtobit) [inline netket/logging/state_log.py:76]:
```suggestion
            mode: Specify the behaviour in case the file already exists at this output_prefix. Options are
                - `[w]rite`: (default) overwrites file/delete the folder if it already exists;
                - `[a]ppend`: appends to the file/folder if it exists, overwise creates a new file;
                - `[x]` or `fail`: fails if file/folder already exists;
            save_every: every how many iterations the variables should be saved. (default: 1)
            tar: if True, state files are saved in a TAR archive instead of a folder. (default: False)
```

File: netket/logging/state_log.py
Comment (femtobit) [inline netket/logging/state_log.py:56]:
```suggestion
    """
        A logger which serializes the variables of the variational state during a run.

        The data is saved either to a directory or tar archive in a sequence of files named
        `[0.mpack, 1.mpack, ...]` where the filename is incremented every time the logger is called.

        The tar file inside is not flushed to disk (closed) until this object is deleted or python
        is shut down.
    """
```

File: netket/logging/state_log.py
Comment (femtobit) [inline netket/logging/state_log.py:66]:
```suggestion
        Initialize the :code:`StateLogger`.
```

**Review by femtobit — APPROVED** [review]
Looks good to me now!

**Comment by PhilipVinc** [comment]
(don't merge yet, I'll try to add a few tests before merging)

**Pull request merged**
